File: cpp/fb_decoder.cpp

#include "schema_generated.h"
#include <fstream>
#include <iostream>

int main(int argc, const char *argv[]) {
    // Read from bin file
    std::ifstream infile;
    infile.open(argv[1], std::ios::binary | std::ios::in);
    infile.seekg(0,std::ios::end);
    int length = infile.tellg();
    infile.seekg(0,std::ios::beg);
    char *data = new char[length];
    infile.read(data, length);
    infile.close();

    auto clients = GetClientList(data);
    auto clients_list = clients->clients();
    for(unsigned int i=0; i < clients_list->size();i++){
        auto client = clients_list->Get(i);
        auto clientType = client->client_type();
        if(clientType == ClientUnion_Person){
            auto person = static_cast<const Person *>(client->client_as_Person());
            auto person_name = person->name();
            auto person_age = person->age();
            auto person_weight = person->weight();
            auto person_gender = person->gender();
            std::cout<<"{ "<<person_name->str()<<","<<person_age<<","<<person_weight<<","<<person_gender->str()<<" }"<<std::endl;
        }else if(clientType == ClientUnion_Group){
            auto group = static_cast<const Group *>(client->client_as_Group());
            auto groupname = group->groupname();
            auto average_age = group->average_age();
            auto average_weight = group->average_weight();
            
            std::cout<<"{ "<<groupname->str()<<","<<average_age<<","<<average_weight<<",";
            auto person_names = group->names();
            std::cout<<" {";
            for(unsigned int j=0; j<person_names->size()-1; j++){
                std::cout<<person_names->Get(j)->str()<<",";
            }
            std::cout<<person_names->Get(person_names->size()-1)->str()<<"}"<<" }";
        }else{
            std::cout<<"Client must be either Person or Group"<<std::endl;
        }
        
        
    }
    return 0;
}

File: py/fb_decoder.py

import flatbuffers
import sys

import Person
import Group
import Client
import ClientList
from ClientUnion import ClientUnion

# Read from bin file
buf = open(sys.argv[1],'rb').read()

client_list = ClientList.ClientList.GetRootAs(buf,0)

number_of_clients = client_list.ClientsLength()

#Iterate on Client_list
for i in range(number_of_clients):
    client = client_list.Clients(i)
    clientType = client.ClientType();
    if clientType == ClientUnion.Person:
        person = Person.Person()
        person.Init(client.Client().Bytes,client.Client().Pos)
        person_name = person.Name()
        person_age = person.Age()
        person_weight = person.Weight()
        person_gender = person.Gender()
        print("{",person_name.decode("utf-8"),",",person_age,",",person_weight,",",person_gender.decode("utf-8"),"}")
    elif clientType == ClientUnion.Group:
        group = Group.Group()
        group.Init(client.Client().Bytes,client.Client().Pos)
        groupname = group.Groupname()
        average_age = group.AverageAge()
        average_weight = group.AverageWeight()
        number_of_peoples = group.NamesLength()
        person_names = "{"
        for j in range(number_of_peoples):
            personName = group.Names(j)
            person_names += personName.decode("utf-8") + ","
        person_names = person_names[:-1] + "}"
        print("{",groupname.decode("utf-8"),",",average_age,",",average_weight,",",person_names,"}")
    else:
        print("Client must be either Person or Group")